flink-libraries/flink-table/src/main/scala/org/apache/flink/table/runtime/aggregate/DataSetSessionWindowAggregateReduceGroupFunction.scala

```diff
@@ -0,0 +1,164 @@
+
+  override def open(config: Configuration) {
+    Preconditions.checkNotNull(aggregates)
+    Preconditions.checkNotNull(groupKeysMapping)
+    aggregateBuffer = new Row(intermediateRowArity)
+    intermediateRowWindowStartPos = intermediateRowArity - 2
+    intermediateRowWindowEndPos = intermediateRowArity - 1
+    output = new Row(finalRowArity)
+    collector = new TimeWindowPropertyCollector(finalRowWindowStartPos, finalRowWindowEndPos)
+  }
+
+  /**
+    * For grouped intermediate aggregate Rows, divide window according to the window-start
+    * and window-end, merge data (within a unified window) into an aggregate buffer, calculate
+    * aggregated values output from aggregate buffer, and then set them into output
+    * Row based on the mapping relationship between intermediate aggregate data and output data.
+    *
+    * @param records Grouped intermediate aggregate Rows iterator.
+    * @param out     The collector to hand results to.
+    *
+    */
+  override def reduce(records: Iterable[Row], out: Collector[Row]): Unit = {
+
+    var last: Row = null
+    var head: Row = null
+    var lastWindowEnd: Option[Long] = None
+    var currentWindowStart: Option[Long] = None
+
+    records.foreach(
+      (record) => {
+        currentWindowStart =
+          Some(record.getField(intermediateRowWindowStartPos).asInstanceOf[Long])
+        // initial traversal or opening a new window
+        if (lastWindowEnd.isEmpty ||
+          (lastWindowEnd.isDefined && currentWindowStart.get > lastWindowEnd.get)) {
+
+          // calculate the current window and open a new window
+          if (lastWindowEnd.isDefined) {
+
+            // evaluate and emit the current window's result.
+            doEvaluateAndCollect(out, last, head)
+          }
+          // initiate intermediate aggregate value.
+          aggregates.foreach(_.initiate(aggregateBuffer))
+          head = record
+        }
+
+        aggregates.foreach(_.merge(record, aggregateBuffer))
+        last = record
+        lastWindowEnd = Some(getWindowEnd(last))
+      })
+
+    doEvaluateAndCollect(out, last, head)
+
+  }
+
+  def doEvaluateAndCollect(
+    out: Collector[Row],
+    last: Row,
+    head: Row): Unit = {
+    // set group keys value to final output.
+    groupKeysMapping.foreach {
+      case (after, previous) =>
+        output.setField(after, last.getField(previous))
+    }
+
+    // evaluate final aggregate value and set to output.
+    aggregateMapping.foreach {
+      case (after, previous) =>
+        output.setField(after, aggregates(previous).evaluate(aggregateBuffer))
+    }
+
+    // adds TimeWindow properties to output then emit output
+    if (finalRowWindowStartPos.isDefined || finalRowWindowEndPos.isDefined) {
+      val start =
+        head.getField(intermediateRowWindowStartPos).asInstanceOf[Long]
+      val end = getWindowEnd(last)
+
+      collector.wrappedCollector = out
+      collector.timeWindow = new TimeWindow(start, end)
```

Comment:
**You are right, maybe we can defined `TimeWindowPropertyCollector`  as follow:**
class TimeWindowPropertyCollector(xxx) extends Collector[Row] {
      ...
     // var timeWindow: TimeWindow = _ 
     var windowStart:Long = _
     var windowEnd:Long = _
   ...
}
**What do you think ?**